images/bridge2-session/tracks/tracks.go

```diff
@@ -0,0 +1,532 @@
+	// these options down. So since we customize the encoding, calling
+	// cbor.Marshal() would lose the options set in the main package.
+	opts := cbor.CoreDetEncOptions()
+	opts.Time = cbor.TimeRFC3339
+	em, err := opts.EncMode()
+	if err != nil {
+		panic(err)
+	}
+	return em
+}()
+
+type Timestamp time.Duration // relative to stream start
+
+func (t Timestamp) String() string {
+	return time.Duration(t).String()
+}
+
+func (t Timestamp) Sub(o Timestamp) time.Duration {
+	return time.Duration(t) - (time.Duration(o))
+}
+
+func (t Timestamp) Add(d time.Duration) Timestamp {
+	return t + Timestamp(d)
+}
+
+type ID string
+
+func (id ID) Compare(other ID) int {
+	// since these are based on xid, we can compare bytes to get time-based ordering
+	return bytes.Compare([]byte(id), []byte(other))
+}
+
+type Span interface {
+	Track() *Track
+	Span(from, to Timestamp) Span
+	Start() Timestamp
+	End() Timestamp
+	Length() time.Duration
+	Audio() beep.Streamer
+	EventTypes() []string
+	Events(typ string) []Event
+}
+
+type Handler interface {
+	HandleEvent(Event)
+}
+
+func newIDWithTime(t time.Time) ID {
+	return ID(xid.NewWithTime(t).String())
+}
+
+func NewID() ID {
+	return newIDWithTime(time.Now().UTC())
+}
+
+type EventMeta struct {
+	Start, End Timestamp
+	Type       string
+	ID         ID
+}
+
+type Event struct {
+	EventMeta
+	Data  any
+	track *Track
+}
+
+func (e Event) Track() *Track {
+	return e.track
+}
+
+func (e Event) Span() Span {
+	return &filteredSpan{e.Start, e.End, e.track}
+}
+
+func (e *Event) UnmarshalCBOR(data []byte) error {
+	type EventRawData struct {
+		EventMeta
+		Data cbor.RawMessage
+	}
+	var eraw EventRawData
+	if err := cbor.Unmarshal(data, &eraw); err != nil {
+		return err
+	}
+	typ, ok := eventTypes[eraw.Type]
+	if !ok {
+		return fmt.Errorf("unknown event type %q", eraw.Type)
+	}
+	value := reflect.New(typ)
+	if err := cbor.Unmarshal(eraw.Data, value.Interface()); err != nil {
+		return err
+	}
+	e.EventMeta = eraw.EventMeta
+	e.Data = reflect.Indirect(value).Interface()
+	return nil
+}
+
+type SessionInfo struct {
+	ID    ID
+	Start time.Time
+}
+
+func LoadSessionInfo(root, id string) (*SessionInfo, error) {
+	x, err := xid.FromString(id)
+	if err != nil {
+		return nil, err
+	}
+	return &SessionInfo{
+		ID:    ID(id),
+		Start: x.Time(),
+	}, nil
+}
+
+var ErrSessionNotFound = fmt.Errorf("session not found")
+
+func LoadSession(root string) (*Session, error) {
+	var s Session
+	f, err := os.Open(filepath.Join(root, "session"))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrSessionNotFound
+		}
+		return nil, err
+	}
+	if err := cbor.NewDecoder(f).Decode(&s); err != nil {
+		return nil, err
+	}
+	// FIXME load the audio tracks as well
+	return &s, nil
+}
+
+type Session struct {
+	EventEmitter
+	ID     ID
+	Start  time.Time
+	tracks sync.Map
+
+	// TODO should we move this logic directly into EventEmitter?
+	EventHandlers []Handler
```

Comment:
I would probably have approached this by making `SessionSaver` responsible for the lifecycle of the session rather than embedding it in the session itself (and renaming it to be more general).
So, there it can provide `Initialize` to load or create a session, register handlers, as well as saving updates.

One challenge I've noticed with the Tractor component system is that since it's all implicit, it can be hard to tell which types are expected to be used as a component. So, when you start nesting component methods too deeply it can be hard to tell which types are meant as components, and which are just meant as standard Go types.

Reply:
My changes were a bit hack and slash... I was trying to balance keeping the diff small and also minimizing the chance of breaking something unexpected. That said, your suggested modification seems reasonable.

I agree, it's hard to know what interfaces are "expected". At some point, it might be interesting to add a sort of go:generate-based interface type assertion.